assets/js/src/integrations/blocks.js

```diff
@@ -1,7 +1,206 @@
+		const item = context?.woocommerce?.cart?.item || context?.item;
+		if ( item ) {
+			return item;
+		}
+	} catch ( e ) {
+		// Invalid JSON, ignore
+	}
+
+	return null;
+};
+
+/**
+ * Extract product data from DOM elements in a cart item row.
+ * This is a fallback when other data sources are unavailable.
+ *
+ * @param {Element} cartItem - The cart item row element.
+ * @return {Object|null} Product object or null if extraction failed.
+ */
+const getProductFromDOM = ( cartItem ) => {
+	const productLink = cartItem.querySelector(
+		'.wc-block-components-product-name'
+	);
+	const productName = productLink?.textContent?.trim();
+
+	if ( ! productName ) {
+		return null;
+	}
+
+	// Try to get quantity from the quantity input or display
+	const quantityInput = cartItem.querySelector(
+		'.wc-block-components-quantity-selector__input'
+	);
+	const quantity = quantityInput
+		? parseInt( quantityInput.value, 10 ) || 1
+		: 1;
+
+	// Try to get price - look for the sale price first, then regular price
+	const priceElement =
+		cartItem.querySelector(
+			'.wc-block-components-product-price__value ins .woocommerce-Price-amount'
+		) ||
+		cartItem.querySelector(
+			'.wc-block-components-product-price__value .woocommerce-Price-amount'
+		) ||
+		cartItem.querySelector( '.wc-block-components-product-price__value' );
+
+	let price = 0;
+	if ( priceElement ) {
+		const priceText = priceElement.textContent
+			?.replace( /[^0-9.,]/g, '' )
+			.replace( ',', '.' );
+		price = parseFloat( priceText ) || 0;
+	}
+
+	// Get currency minor unit (decimal places) - default to 2
+	const currencyMinorUnit = 2;
+
+	// Build a minimal product object that works with getProductFieldObject
+	return {
+		name: productName,
+		quantity,
+		prices: {
+			price: Math.round( price * 10 ** currencyMinorUnit ),
+			currency_minor_unit: currencyMinorUnit,
+		},
+	};
+};
+
+/**
+ * Track when an item is removed from the Interactivity API-powered Mini Cart.
+ * The new Mini Cart (WooCommerce 10.4+) doesn't fire the experimental__woocommerce_blocks
+ * cart-remove-item action, so we need to listen for clicks on the remove button.
+ *
+ * This listener only fires if the hook hasn't already handled the removal to prevent
+ * duplicate events on older WooCommerce versions.
+ *
+ * @param {Function} getEventHandler - Function to get the event handler for a given event name.
+ */
+const trackMiniCartRemoval = ( getEventHandler ) => {
+	document.body.addEventListener( 'click', ( event ) => {
+		const removeButton = event.target.closest(
+			'.wc-block-cart-item__remove-link'
+		);
+
+		if ( ! removeButton ) {
+			return;
+		}
+
+		// Find the cart item container to get product information
+		const cartItem = removeButton.closest( '.wc-block-cart-items__row' );
+		if ( ! cartItem ) {
+			return;
+		}
+
+		let product = null;
+
+		// 1. Try Interactivity API context first (WooCommerce 10.4+)
+		product = getProductFromInteractivityContext( cartItem );
```

Comment:
Could you share what conditions make it possible to get the product via `getProductFromInteractivityContext`? I've tried several different combinations of pages and themes, but haven't found it.